doc/architectural/central-data-structures.md

```diff
@@ -0,0 +1,99 @@
+# Central Data Structures
+
+The following is some light technical documentation of the major data structures
+represented in the input transformation to Intermediate Representation (IR) inside
+CBMC and the assorted CProver tools.
+
+## goto_modelt
+
+The `goto_modelt` is the main data structure that CBMC (and the other tools) use to
+represent GOTO-IR (the `GOTO` Intermediate Representation).
+
+A `goto_modelt` is effectively a pair, consisting of:
+
+* A list of GOTO functions,
+* A symbol table containing symbol references for the symbols contained in the GOTO functions.
+
+In pseudocode, the type looks this:
+
+```js
+type goto_modelt {
+  type goto_functionst = list<goto_functiont>
+  type symbol_tablet = map<identifier, symbolt>
+}
+```
+
+The abstract interface of `goto_modelt` is outlined in the file
+[`src/goto-programs/abstract_goto_model.h`](../../src/goto-programs/abstract_goto_model.h).
+
+## goto_functiont
+
+A `goto_functiont` is also a product type. It's designed to represent a function
+at the IR level, and effectively it's the following ADT (in pseudocode):
+
+```js
+type goto_functiont {
+    goto_programt body
+    list<identifiers> parameters
+}
+```
+
+Of the two types listed above, the `parameters` one should be self-documenting,
+(the values of these are later looked up in the symbol table), so we're going to
+focus next on the type `goto_programt`
+
+## goto_programt
+
+A `goto_programt` is a container of GOTO-IR instructions. In pseudocode, it would
+look like `type goto_programt = list<goto_instructiont>`.
+
+An instruction (`goto_instructiont`) is another product type, describing a GOTO level
+instruction with the following 3 component subtypes, again in pseudocode:
+
+```js
+type goto_instructiont {
+    instruction_type instr_type
+    instruction      statement
+    guard            boolean_expr
+}
+```
+
+The above subtypes are just illustrative, so we will provide some extra explanation for
+these:
+
+* The `instruction_type` above corresponds to the `goto_program_instruction_typet` type
+  listed in [`goto_program.h`](../../src/goto-programs/goto_program.h)
+  * For illustration purposes, some instruction types are `assign`, `function_call`, `return`,
+    etc.
+* The `instruction` is a statement represented by a `goto_instruction_codet`.
+  * A `goto_instruction_codet` is a `codet` (a data structure broadly representing a statement
+    inside CBMC) that contains the actual GOTO-IR instruction.
```

Comment:
So a GOTO-IR `modelt` contains GOTO-IR `functiont`s which contains a list of  GOTO-IR `body`s which are GOTO-IR `programt`s that contain **GOTO-IR `instructiont`s** that contain an `instruction` that contains _GOTO-IR `instruction_codet`_ that contain **_GOTO-IR `codet`_** that contain _GOTO-IR `instruction_codet`_ that is a **_GOTO-IR `codet`_** that contains **GOTO-IR `instructiont`**... and this is not at all confusing or circular?

Or have I misread the documentation to this point?

Reply:
I didn't choose the class names 😅

Yes, your understanding is more or less correct. I'll see if I can simplify it further, but I can't promise massive improvements - this is just the description of the status quo.

Reply:
Some of the type names involved seem a little obtuse, which is just why it is so important to document them. Some of the relationships as expressed by @TGWDB are not quite right however. Given that this has been misunderstood by one person already, it suggests that we haven't managed to explain the relationships sufficiently clearly in this documentation.

- A model **contains** a collection of functions.
- Each function **contains** a single "body" which **is** of type `goto_programt`.
- A "body" **contains** an ordered collection of `instructiont`.
- An `instructiont` may optionally **contain** a "code" expression which **is** of type `codet`. `codet` and `goto_instruction_codet` are currently defined to be the same type, but this is not guaranteed to be the case in future.

Note that a function has a single "body" not a collection of them. The fact that a function body is of type `goto_programt` is the especially confusing point, as it is not an entire program! The set of data structures are not circular. There are 4 container ship relations between 5 levels of data.

The `instructiont` class type works a bit like a tagged union. For some values of the instruction type, the `code` member is constrained to be a particular sub-class of `codet`. For others such as the `SKIP` instruction type, the `code` member is expected to be empty (is `ID_nil`). Given that instructions also have other data members `intructiont` is both a sum and product type.